### PR TITLE
feat: universal router sdk integration

### PR DESCRIPTION
**Pull request by ChefBingbong**
### Summary
🥞🌐🎨

This pull request adds the `@pancakeswap/universal-router-sdk` package that enables the swap feature to support multiple chains and protocols for swapping NFTs and ERC20s using the Universal Router contracts. It also updates the `useSwapCallArguments` function and the `swapCallParameters` test to use the new package. It also modifies some configuration and utility files for testing and development purposes.

> _We are the universal routers, we swap across the chains_
> _We use the `@pancakeswap/universal-router-sdk`, we have no restraints_
> _We pool and zap the NFTs, we trade with ERC20s_
> _We are the universal routers, we defy the protocols' domains_

### Walkthrough
*  Add `@pancakeswap/universal-router-sdk` package to interact with Universal Router contracts ([link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-197cd8ca285a4abd2f21479e0bf6e36e90b08528fcd7f3bdbe8d1221897e377dR60), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8R45), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L14-R14), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-35b64341bb3c5ad954125e648cf61d7dafa963042e235085982f678bea479452R1-R8), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-0191b24758eea2a25c5d112333cb917b6023dd65ab90df1695667b0a2c94a1fbR1-R3), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-ab28a5ee787d023b6d2a912d77c996d8a9de058056af77825d92ed25193edffbR1-R102), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-f5eea6ca20f65d81626764a5684bd0b1c3662adb71743ba8677384d0002f51ebR1-R15), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-f6540822464dc0df575ca5e40418c32c63bce949b2fdb8fced51c4be7d3995f7R1-R80), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-3bd8eeac545c378152790612294e83d0506314a367dab279e7a2b71015c5c065R1-R104), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-1f160ccae2e1c481dc2510484778fea766ca096c9f75d80e45b61d51e19c0ee2L1-R221), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-087fd51a2692b8c57faaa88777edd19347c9033539fdb3929fe77c992746f8dfR1-R242), [link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-2eaf865d1d5b645ba8da6c0be620114b0276c1f71c225f924191652c23abdd58R1-R250))
*  Replace `SwapRouter.swapCallParameters` with `PancakeUniSwapRouter.swapERC20CallParameters` to use custom method for swapping ERC20s ([link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L52-R52))
*  Replace `swapRouterAddress` with `UNIVERSAL_ROUTER_ADDRESS` to use function that returns correct contract address ([link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L100-R100))
*  Import `PancakeUniSwapRouter` and `UNIVERSAL_ROUTER_ADDRESS` from `@pancakeswap/universal-router-sdk` in `useSwapCallArguments.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L9-R9))
*  Modify `swapCallParameters` test to use mixed route trades in `swapRouter.test.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-d48ae241875ea680e3d23f0cd3e29fa944ebe38a0f7c7b9731ab913f8a0314efL428-R428))
*  Change `basisPointsToPercent` function to return fixed percent in `exchange.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7847/files?diff=unified&w=0#diff-239464a331297352861e462676de18b620775aef2c455102ab85198f4f333e06L21-R21))

### CODE REVIEW

File: apps/web/next.config.mjs [hunk 1]
```diff
@@ -57,6 +57,7 @@ const config = {
     '@pancakeswap/hooks',
     '@pancakeswap/utils',
     '@pancakeswap/widgets-internal',
+    '@pancakeswap/universal-router-sdk'
```
Comment (chefjackson) [inline apps/web/next.config.mjs:60]: Let's publish our sdk before merge. Remember to remove the transpile after publish

File: apps/web/package.json [hunk 2]
```diff
@@ -47,14 +47,16 @@
...
     "@pancakeswap/utils": "workspace:*",
     "@pancakeswap/v3-sdk": "workspace:*",
     "@pancakeswap/wagmi": "workspace:*",
     "@pancakeswap/widgets-internal": "workspace:*",
     "@reduxjs/toolkit": "^1.9.1",
-    "@tanstack/react-query": "^4.28.0",
     "@sentry/nextjs": "^7.52.1",
     "@snapshot-labs/snapshot.js": "^0.4.40",
+    "@tanstack/react-query": "^4.28.0",
+    "@uniswap/permit2-sdk": "^1.2.0",
```
Comment (chefjackson) [inline apps/web/package.json:59]: Let's setup our own permit2 sdk as well

File: apps/web/src/config/abi/permit2.json [hunk 3]
```diff
@@ -0,0 +1,42 @@
+[
```
Comment (chefjackson) [inline apps/web/src/config/abi/permit2.json:1]: let's move the abi to our permit2 sdk package. And let's use ts file to export the abi for better typing

File: apps/web/src/hooks/usePermit2Allowance.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+      })
+    }
+  }, [isApprovalPending])
+
+  // Signature and PermitAllowance will expire, so they should be rechecked at an interval.
+  // Calculate now such that the signature will still be valid for the submitting block.
+  const [now, setNow] = useState(Date.now() + AVERAGE_L1_BLOCK_TIME)
+  useInterval(
+    useCallback(() => setNow((Date.now() + AVERAGE_L1_BLOCK_TIME) / 1000), []),
+    AVERAGE_L1_BLOCK_TIME
```
Comment (chefjackson) [inline apps/web/src/hooks/usePermit2Allowance.ts:91]: block time varies for different chain

File: packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+        planner.addCommand(CommandType.PAY_PORTION, [
+          sampleTrade.outputAmount.currency.wrapped.address,
+          this.options.fee.recipient,
+          feeBips,
+        ])
+
+        // If the trade is exact output, and a fee was taken, we must adjust the amount out to be the amount after the fee
+        // Otherwise we continue as expected with the trade's normal expected output
+        if (this.type === TradeType.EXACT_OUTPUT) {
+          minAmountOut = minAmountOut.subtract(minAmountOut.multiply(feeBips))
```
Comment (ChefJerry) [inline packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts:121]: `minAmountOut` is not a BigNumber now, no `subtract` method

File: apps/web/package.json [hunk 6]
```diff
@@ -4,7 +4,7 @@
   "private": true,
   "scripts": {
     "analyze": "ANALYZE=true pnpm build",
-    "dev": "next dev",
+    "dev": "next dev -p 3001",
```
Comment (chefjackson) [inline apps/web/package.json:7]: let's remove the port here

File: apps/web/src/config/abi/permit2.json
Comment (chefjackson) [inline apps/web/src/config/abi/permit2.json:1]: no need for this file right? Can import from the permit2 sdk

File: apps/web/src/hooks/useApproveCallback.ts
Comment (chefjackson) [inline apps/web/src/hooks/useApproveCallback.ts:1]: can we leave this file untouched?

File: apps/web/src/hooks/useCallWithGasPrice.ts [hunk 7]
```diff
@@ -94,7 +94,8 @@ export function useCallWithGasPrice() {
         functionName: methodName,
         args: methodArgs,
         gasPrice,
-        gas: calculateGasMargin(gas),
+        // for some reason gas price is insamely high when using maxuint approval, so commenting out for now
+        // gas: calculateGasMargin(gas),
```
Comment (chefjackson) [inline apps/web/src/hooks/useCallWithGasPrice.ts:98]: if the gas estimation is incorrect, probably something wrong with the tx itself. Let's add it back and find the root cause

**Pull request closed without merging**